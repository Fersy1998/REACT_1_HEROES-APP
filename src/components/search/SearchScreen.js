import React, { useMemo } from 'react'
import { Navigate, useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm'
import { getHeroByName } from '../../selectors/getHeroByName';
import { HeroCard } from '../hero/HeroCard';
import queryString from 'query-string'
export const SearchScreen = () => {
    const navegador=useNavigate();
    const location=useLocation();
    //const query=queryString.parse(location);
    //Si q no existe entonces es vacío
    const {q=''}=queryString.parse(location.search);
    const [formValues, handleInputChange] = useForm({searchElement:q});
    const {searchElement}=formValues;
    const HeroesFiltered= useMemo(()=>getHeroByName(q), [q]);
    const handleSubmit=(e)=>{
        e.preventDefault();
        navegador(`?q=${searchElement}`);
        //console.log(HeroesFiltered);
    }
    
    
  return (
    <div>
        <h1 >Búsqueda</h1> <hr />
        <div className='row'>
            <div className='col-5'></div>
                <form onSubmit={handleSubmit}>
                    <input type='text' 
                    placeholder='buscar un héroe'
                    className='form-control'
                    name='searchElement'
                    autoComplete='off'
                    value={searchElement}
                    onChange={handleInputChange}
                    />
                    <button className='btn btn-primary mt-5'type='submit' onClick={handleSubmit}>Buscar...</button>
                </form>
        
            </div>
            <div className='col-7 w-100 h-100'>
                <h4>Resultados</h4><hr />
                <div className='row'>
                {
                    (q==='')
                        ? <div className='alert alert-info'>Buscar un héroe</div>
                        :(HeroesFiltered.length===0)
                        &&  <div className='alert alert-danger'>Sin resultados para la búsqueda {q}</div>
                }
                
                {
                   HeroesFiltered.map(hero=>(<div className='col col-3 animate__animated animate__backInLeft' key={hero.id}><HeroCard hero={hero} /></div>))
                
                }
                </div>
            </div>
    </div>
  )
}
