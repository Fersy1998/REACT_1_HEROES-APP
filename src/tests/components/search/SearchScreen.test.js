import { SearchScreen } from "../../../components/search/SearchScreen";
import {mount} from 'enzyme'
import { MemoryRouter } from "react-router-dom";


//Hacems mock de react-router-dom, pero copiamos todos los demas elementos y los dejamos igual
//y modificamos solamente el useNavigate para poderlo usar como mock en la prueba 
// de -->should call navigate to a new Screen
//Debe llamarse mockNavigate para que funcione
const mockNavigate=jest.fn();

jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate:()=> mockNavigate,
}))

describe('Pruebas en SearhcScreen.js', ()=>{
    test('should shows corrrectly with default values', () => { 
        const wrapper=mount(
            <MemoryRouter initialEntries={['/search']}>
            
                <SearchScreen />
            </MemoryRouter>
        
        
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text()).toBe('Buscar un héroe');
    
    })
    test('should Batman and the input with queryString value', () => { 
        const wrapper=mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
            
                <SearchScreen />
            </MemoryRouter>)
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })
    test('should show an error if hero isn´t found', () => { 
        const wrapper=mount(
            <MemoryRouter initialEntries={['/search?q=batman34535']}>
            
                <SearchScreen />
            </MemoryRouter>)
        expect(wrapper.find('.alert-danger').text()).toEqual(`Sin resultados para la búsqueda batman34535`);
        expect(wrapper).toMatchSnapshot();
    
    })
    test('should call navigate to a new Screen', () => { 
        const wrapper=mount(
            <MemoryRouter initialEntries={['/search']}>
            
                <SearchScreen />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change', {target:{
            name:'searchElement',
            value:'batman'
        }})
        wrapper.find('form').prop('onSubmit')({ preventDefault(){}})
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman')
    })

})