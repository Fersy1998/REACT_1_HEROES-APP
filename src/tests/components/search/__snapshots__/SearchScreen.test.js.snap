// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearhcScreen.js should Batman and the input with queryString value 1`] = `
<div>
  <h1>
    Búsqueda
  </h1>
   
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchElement"
        onChange={[Function]}
        placeholder="buscar un héroe"
        type="text"
        value="batman"
      />
      <button
        className="btn btn-primary mt-5"
        onClick={[Function]}
        type="submit"
      >
        Buscar...
      </button>
    </form>
  </div>
  <div
    className="col-7 w-100 h-100"
  >
    <h4>
      Resultados
    </h4>
    <hr />
    <div
      className="row"
    >
      <div
        className="col col-3 animate__animated animate__backInLeft"
        key="dc-batman"
      >
        <div
          className="card  "
          style={
            Object {
              "width": "auto",
            }
          }
        >
          <img
            alt="Batman"
            className="card-img-top"
            src="/assets/dc-batman.jpg"
          />
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Detective Comics #27
            </p>
            <p
              className="card-text"
            >
              DC Comics
            </p>
            <p
              className="card-text"
            >
              Bruce Wayne
            </p>
            <a
              className="btn btn-primary"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Ver más
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearhcScreen.js should show an error if hero isn´t found 1`] = `
<div>
  <h1>
    Búsqueda
  </h1>
   
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchElement"
        onChange={[Function]}
        placeholder="buscar un héroe"
        type="text"
        value="batman34535"
      />
      <button
        className="btn btn-primary mt-5"
        onClick={[Function]}
        type="submit"
      >
        Buscar...
      </button>
    </form>
  </div>
  <div
    className="col-7 w-100 h-100"
  >
    <h4>
      Resultados
    </h4>
    <hr />
    <div
      className="row"
    >
      <div
        className="alert alert-danger"
      >
        Sin resultados para la búsqueda 
        batman34535
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearhcScreen.js should shows corrrectly with default values 1`] = `
<div>
  <h1>
    Búsqueda
  </h1>
   
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchElement"
        onChange={[Function]}
        placeholder="buscar un héroe"
        type="text"
        value=""
      />
      <button
        className="btn btn-primary mt-5"
        onClick={[Function]}
        type="submit"
      >
        Buscar...
      </button>
    </form>
  </div>
  <div
    className="col-7 w-100 h-100"
  >
    <h4>
      Resultados
    </h4>
    <hr />
    <div
      className="row"
    >
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>
</div>
`;
